x/superfluid/keeper/migrate.go

```diff
@@ -1,99 +1,180 @@
+		sharesToMigrate = gammSharesInLock
+	}
+
+	// Otherwise, we must ensure that the shares to migrate is less than or equal to the shares in the lock.
+	if sharesToMigrate.Amount.GT(gammSharesInLock.Amount) {
+		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, fmt.Errorf("shares to migrate must be less than or equal to shares in lock")
+	}
+
+	// Exit the balancer pool position.
+	exitCoins, err := k.gk.ExitPool(ctx, sender, poolIdLeaving, sharesToMigrate.Amount, sdk.NewCoins())
+	if err != nil {
+		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, err
+	}
+	// Defense in depth, ensuring we are returning exactly two coins.
+	if len(exitCoins) != 2 {
+		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, fmt.Errorf("Balancer pool must have exactly two tokens")
+	}
+
+	// Create a full range (min to max tick) concentrated liquidity position.
+	if wasSuperfluidDelegatedBeforeMigration {
+		// If the lock was previously superfluid delegated, we create a new lock, keep it locked, and superfluid delegate it.
+		positionId, amount0, amount1, liquidity, joinTime, concentratedLockId, err = k.clk.CreateFullRangePositionLocked(ctx, concentratedPool, sender, exitCoins, remainingLockTime)
+		if err != nil {
+			return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, err
+		}
+		err := k.SuperfluidDelegate(ctx, sender.String(), concentratedLockId, intermediateAccount.ValAddr)
+		if err != nil {
+			return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, err
+		}
+	} else {
+		// If the lock was unlocking, we create a new lock that is unlocking for the remaining time of the old lock.
+		positionId, amount0, amount1, liquidity, joinTime, concentratedLockId, err = k.clk.CreateFullRangePositionUnlocking(ctx, concentratedPool, sender, exitCoins, remainingLockTime)
+		if err != nil {
+			return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, err
+		}
+	}
+
+	// If the lock was superfluid undelegating at time of migration
+	if wasSuperfluidUndelegatingBeforeMigration {
+		// The previous intermediary account is now invalid for the new lock, since the underlying denom has changed and intermediary accounts are
+		// created by validator address, denom, and gauge id.
+		// We must therefore create and set a new intermediary account based on the previous validator but with the new lock's denom.
+		concentratedLockupDenom := cltypes.GetConcentratedLockupDenomFromPoolId(poolIdEntering)
+		clIntermediateAccount, err := k.GetOrCreateIntermediaryAccount(ctx, concentratedLockupDenom, valAddr)
+		if err != nil {
+			return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, err
+		}
+
+		// Create a new synthetic lockup for the new intermediary account in an unlocking status
+		err = k.createSyntheticLockup(ctx, concentratedLockId, clIntermediateAccount, unlockingStatus)
+		if err != nil {
+			return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, err
+		}
+	}
```

Comment:
Why do we not need to create a superfluid position and delegate it prior to running the undelegating logic in this case?

Reply:
Because a undelegating position has no funds delegated to a validator (when we initiate undelegate, we immediately are undelegated and the underlying osmo is burned). The way a undelegating lockup gets slashing is by iterating the unbonding synth locks. See below:
![Screenshot 2023-04-19 at 12 21 09 PM](https://user-images.githubusercontent.com/40078083/233152059-89150288-3679-41d8-a61a-9075e7cc04cd.png)